### DISCUSSION TITLE
Undefined method `default_url_options' when using Component and Phlex::Rails

### DISCUSSION THREAD

**Discussion by LolsonX** [yippee-fun/phlex#767] [Help] [answered]
Hi, i am exploring Phlex with Rails. I wanted to render component inside view, with some logic about paths, but for some reason I get
`Undefined method `default_url_options' for nil` in route helper error

```ruby
class Articles::NewView < ApplicationView
  attr_reader :article

  def initialize(article:)
    @article = article
  end

  def view_template
    h1 { "New article" }
    render Articles::FormComponent.new(article: article)
  end
end
```

The component looks like this:
```ruby
class Articles::FormComponent < ApplicationComponent
  attr_reader :article, :action, :method


  def initialize(article:)
    @article = article
    if article.persisted?
      @action = article_path(article) # Here
      @method = :patch
    else
      @action = articles_path() # and here
      @method = :post
    end
  end
  
  def view_template
    # template content...
  end
end
```

My ApplciationComponent looks like this:
```ruby
class ApplicationComponent < Phlex::HTML
  include Phlex::Rails::Helpers::Routes
  include Phlex::Rails::Helpers::LinkTo

  if Rails.env.development?
    def before_template
      comment { "Before #{self.class.name}" }
      super
    end
  end
end
```

What's interesting defining 2 methods allowed me to render view
```ruby
class Articles::FormComponent < ApplicationComponent
  def url_options
    {}
  end

  def default_url_options
    {}
  end
end
 ```
 
 I am wondering is it a potential bug or just my poor skills :D

**Comment by joeldrapper** [comment]
That’s very strange. Should be handled by the inclusion of the `Routes` adapter. It sounds like it’s getting to that adapter but then it’s calling `default_url_options` on `nil`. The adapter should call `default_url_options` on `helpers`, which is an alias for the view context, which should have been passed in the first time Rails handed over to Phlex and then Phlex should have passed it down from parent to child.

I expect something is going wrong way before we get to this helper. How are you rendering `Articles::NewView`? I assume that’s in the controller something like this?

```ruby
render Articles::NewView.new(
  article: Article.new
)
```

**Comment by LolsonX** [comment]
Yeah exactly

**Comment by joeldrapper** [comment, marked as answer]
Oh, I know what it is! We should really call this out in the guides. You can’t use helpers from the initializer, because the initializer happens when the component is created, not when it’s rendered. We don’t have a rendering context at that point.

If we write the controller action like this, it’s more apparent.

```ruby
component = Articles::NewView.new(
  article: Article.new
)

# the initializer has run at this point, but it doesn’t have a rendering context

render(component)

# now it has a rendering context but it’s too late
```

The solution is to move that logic out.

```ruby
class Articles::FormComponent < ApplicationComponent
  attr_reader :article, :action, :method

  def initialize(article:)
    @article = article
  end

  def before_template
    super # it’s good practice to call this from before_template callbacks

    if @article.persisted?
      @action = article_path(@article)
      @method = :patch
    else
      @action = articles_path
      @method = :post
    end
  end
  
  def view_template
    # template content...
  end
end
```

Alternatively, you could calculate these on the fly

```ruby
class Articles::FormComponent < ApplicationComponent
  def initialize(article:)
    @article = article
  end
  
  def view_template
    # template content...
  end

  private

  def action
    @article.persisted? ? article_path(@article) : articles_path
  end

  def method
    @article.persisted? ? :patch : :post
  end
end
```

Unrelated, I recommend avoiding `attr_reader`s as they often conflict with HTML tags. Your example overwrote the method for creating `<article>` tags. I would reference instance variables directly. It’s faster too.

**Reply by willcosgrove** [reply]
Just thought I would call this out for future readers: `before_render` in the example above should be `before_template`.

**Comment by LolsonX** [comment]
Thanks for help and advice, will try out when back home
